### PR TITLE
Build examples with Github Actions

### PR DESCRIPTION
**Pull request by netantho**
## Status
**READY FOR REVIEW**

## Breaking Changes
NO

## Description
Analyze and build the examples (APK, js) with Github Actions to be more confident they are working and if not we'll notice.

## Long term goals

In future PRs, I hope to continue making examples more reliable:
- create tests for the examples
- use [embedded dart pads](https://github.com/dart-lang/dart-pad/wiki/Embedding-Guide) for the docs, so that only built and tested code appears in the doc

## Non goals

As suggested by @tenhobi, building on Mac is not great with Github Actions for now so we're not building iOS apps. I've opened https://github.com/felangel/bloc/issues/835 to reconsider in the future.

## Todos
- packages/angular_bloc/example
  - [x] analyze - Won't fix, delete as it's essentially the same as `example/angular_counter` but not buildable nor runnable.
  - [x] build js - Won't fix, delete as it's essentially the same as `example/angular_counter` but not buildable nor runnable.
- packages/bloc/example/
  - [x] analyze
  - [x] path-based deps to bloc packages
- packages/bloc_test/example/main.dart
  - [x] analyze
  - [x] run tests
  - [x] path-based deps to bloc packages
- packages/flutter_bloc/example/
  - [x] migrate to AndroidX
  - [x] analyze
  - [x] build APK
  - [x] path-based deps to bloc packages
- examples/angular_counter
  - [x] analyze
  - [x] build js
  - [x] path-based deps to bloc packages
- examples/flutter_bloc_with_stream
  - [x] migrate to AndroidX
  - [x] analyze
  - [x] build APK
  - [x] path-based deps to bloc packages
- examples/flutter_complex_list
  - [x] migrate to AndroidX
  - [x] analyze
  - [x] build APK
  - [x] path-based deps to bloc packages
- examples/flutter_dynamic_form
  - [x] migrate to AndroidX
  - [x] analyze
  - [x] build APK
  - [x] path-based deps to bloc packages
- examples/flutter_firebase_login
  - [x] migrate to AndroidX - was already migrated
  - [x] analyze
  - [x] build APK
  - [x] path-based deps to bloc packages
- examples/flutter_firestore_todos
  - [x] migrate to AndroidX - was already migrated
  - [x] analyze
  - [x] build APK
  - [x] path-based deps to bloc packages
- examples/flutter_form_validation
  - [x] migrate to AndroidX
  - [x] analyze
  - [x] build APK
  - [x] path-based deps to bloc packages
- examples/flutter_infinite_list
  - [x] migrate to AndroidX
  - [x] analyze
  - [x] build APK
  - [x] path-based deps to bloc packages
- examples/flutter_login
  - [x] migrate to AndroidX
  - [x] analyze
  - [x] build APK
  - [x] path-based deps to bloc packages
- examples/flutter_shopping_cart
  - [x] migrate to AndroidX
  - [x] analyze
  - [x] build APK
  - [x] path-based deps to bloc packages
- examples/flutter_timer
  - [x] migrate to AndroidX
  - [x] analyze
  - [x] build APK
  - [x] path-based deps to bloc packages
- examples/flutter_todos
  - [x] migrate to AndroidX
  - [x] analyze
  - [x] build APK
  - [x] path-based deps to bloc packages
- examples/flutter_weather
  - [x] migrate to AndroidX
  - [x] analyze
  - [x] build APK
  - [x] path-based deps to bloc packages
- examples/github_search/common_github_search
  - [x] analyze
  - [x] path-based deps to bloc packages
- examples/github_search/angular_github_search
  - [x] analyze
  - [x] build js
  - [x] path-based deps to bloc packages
- examples/github_search/flutter_github_search
  - [x] migrate to AndroidX
  - [x] analyze
  - [x] build APK
  - [x] path-based deps to bloc packages
- performance_tests/flutter_bloc/bloc_builder
  - [x] migrate to AndroidX
  - [x] analyze
  - [x] path-based deps to bloc packages
- performance_tests/flutter_bloc/bloc_provider
  - [x] migrate to AndroidX
  - [x] analyze
  - [x] path-based deps to bloc packages
- [x] Update instructions for creating a new example

## Steps to Test or Reproduce
Outline the steps to test or reproduce the PR here.

```
```

## Impact to Remaining Code Base
This PR will affect:

*

### CODE REVIEW

**Comment by felangel** [comment]
This is amazing! Thanks so much for doing this and let me know when it's ready for review 🚀

**Comment by tenhobi** [comment]
Hi, very nice! I was thinking about the same but didn't have time to do it. :-)

We should consider creating its own action so it's not run every time when nothing changed there. That means copying exactly the `main.yaml`, but also adding path settings. Otherwise, any change would also check those examples, which would take no time + have no impact with not changed settings.

Take a look at https://help.github.com/en/actions/automating-your-workflow-with-github-actions/workflow-syntax-for-github-actions#onpushpull_requestpaths

File: .github/workflows/main.yaml [hunk 1]
```diff
@@ -74,10 +75,50 @@ jobs:
...
       - name: Upload coverage to Codecov
         uses: codecov/codecov-action@v1
         with:
           token: ${{ secrets.CODECOV_TOKEN }}
           file: ${{ matrix.package }}.lcov
+
+  # Build an APK for each flutter example.
+  build-apk:
+    runs-on: ${{ matrix.os }}
+    needs: setup
```
Comment (tenhobi) [inline .github/workflows/main.yaml]: I think this would need the `analyze-and-test` job more as it's time-consuming to run build when tests or something else fails, as the build probably also fails?

**Comment by tenhobi** [comment]
For `Won't fix, needs google-services.json for firebase` we can actually add some example firebase setup (cc @felangel) to either source code as for example `./example_google_services.json` and then in CI move it (if exists) to the actual folder with the required name. So we can build it. ;)

**Comment by tenhobi** [comment]
Also, as discussed in https://github.com/felangel/bloc/pull/827, I would delay building iOS. Mostly because we cannot cache flutter on both systems (2 GB limit) and Flutter is kinda big.

File: scripts/ci.sh [hunk 2]
```diff
@@ -15,8 +15,8 @@ if grep -q 'sdk: flutter' "./pubspec.yaml"; then
       flutter format --set-exit-if-changed test
       flutter analyze --no-current-package test
       flutter test --no-pub --coverage
+      cp ./coverage/lcov.info ../../$package.lcov
```
Comment (tenhobi) [inline scripts/ci.sh]: This would break the regular CI, as it will not go to the if.

**Comment by tenhobi** [comment]
@netantho Nice changes. I would rather use `paths` to **whitelist** a path, not `ignore-paths` to **blacklist** the paths. This way you actually tell what paths should run the CI, not what paths don't, which is more universal.

image

File: .github/workflows/main.yaml (hunk 1 above)
Comment (netantho) [reply .github/workflows/main.yaml]: Great idea, implemented.

File: scripts/ci.sh (hunk 2 above)
Comment (netantho) [reply scripts/ci.sh]: My understanding is that `./coverage/lcov.info` is generated by `flutter test --no-pub --coverage`. So it should work?

**Comment by netantho** [comment]
Thanks @tenhobi for the great suggestions!

> Also, as discussed in #827, I would delay building iOS. Mostly because we cannot cache flutter on both systems (2 GB limit) and Flutter is kinda big.
> 
> Let me know if you need any help. 😉

I opened https://github.com/felangel/bloc/issues/835 to track this.

> Hi, very nice! I was thinking about the same but didn't have time to do it. :-)
> 
> We should consider creating its own action (`examples.yaml`) so it's not run every time when nothing changed there. That means copying exactly the `main.yaml`, but also adding path settings. Otherwise, any change would also check those examples, which would take no time + have no impact with not changed settings.
> 
> Take a look at https://help.github.com/en/actions/automating-your-workflow-with-github-actions/workflow-syntax-for-github-actions#onpushpull_requestpaths
> 
> Edit: And we could then make separate badge `examples: passing`. 😄

Great idea, implemented.

> @netantho Nice changes. I would rather use `paths` to **whitelist** a path, not `ignore-paths` to **blacklist** the paths. This way you actually tell what paths should run the CI, not what paths don't, which is more universal.
> 
> image

Done

File: scripts/ci.sh (hunk 2 above)
Comment (tenhobi) [reply scripts/ci.sh]: Oh, I see.

**Comment by netantho** [comment]
> For `Won't fix, needs google-services.json for firebase` we can actually add some example firebase setup (cc @felangel) to either source code as for example `./example_google_services.json` and then in CI move it (if exists) to the actual folder with the required name. So we can build it. ;)

Great idea 👍 done 🚀

File: .github/workflows/examples.yaml [hunk 3]
```diff
@@ -71,7 +71,7 @@ jobs:
     strategy:
       matrix:
         os: [ubuntu-latest]
-        folder: ['packages/angular_bloc/example']
+        folder: ['packages/bloc/example', 'package/bloc_test/example', 'examples/angular_counter']
```
Comment (tenhobi) [inline .github/workflows/examples.yaml]: Dont forget to include and exclude `packages/*/example` to paths in both CI files :-)

Comment (tenhobi) [reply .github/workflows/examples.yaml]: like this:

```
on:
  push:
    paths:
    - 'sub-project/**'
    - '!sub-project/docs/**'
```

https://help.github.com/en/actions/automating-your-workflow-with-github-actions/workflow-syntax-for-github-actions#example-using-positive-and-negative-patterns-1

File: packages/angular_bloc/example/example.dart [hunk 4]
```diff
@@ -1,69 +0,0 @@
-import 'dart:async';
```
Comment (felangel) [inline packages/angular_bloc/example/example.dart]: I don't think we should delete this file because it negatively impacts the pub score and is [displayed directly on pub.dev](https://pub.dev/packages/angular_bloc#-example-tab-)

File: scripts/ci_build_js.sh [hunk 5]
```diff
@@ -0,0 +1,13 @@
+#!/bin/bash
+
+set -ex
+
+source scripts/retry.sh
```
Comment (felangel) [inline scripts/ci_build_js.sh]: do we need this?

File: scripts/ci_build_apk.sh [hunk 6]
```diff
@@ -0,0 +1,17 @@
+#!/bin/bash
+
+set -ex
+
+source scripts/retry.sh
```
Comment (felangel) [inline scripts/ci_build_apk.sh]: do we need this?

File: packages/bloc_test/pubspec.yaml [hunk 7]
```diff
@@ -9,7 +9,8 @@ environment:
   sdk: ">=2.6.0 <3.0.0"
 
 dependencies:
-  bloc: ^3.0.0
+  bloc:
```
Comment (felangel) [inline packages/bloc_test/pubspec.yaml]: why is this change necessary? I'd prefer to keep it as is since we cannot publish with path imports

File: packages/angular_bloc/pubspec.yaml [hunk 8]
```diff
@@ -10,7 +10,8 @@ environment:
 
 dependencies:
   angular: ^5.3.0
-  bloc: ^3.0.0
+  bloc:
```
Comment (felangel) [inline packages/angular_bloc/pubspec.yaml]: same comment here about why this is needed?

File: examples/github_search/flutter_github_search/pubspec.yaml [hunk 9]
```diff
@@ -9,8 +9,9 @@ environment:
 dependencies:
   flutter:
     sdk: flutter
-  flutter_bloc: ^3.0.0
-  url_launcher: ^4.0.3
+  flutter_bloc:
```
Comment (felangel) [inline examples/github_search/flutter_github_search/pubspec.yaml]: same comment here about why we need to change to path?

File: .github/workflows/examples.yaml [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+    - 'examples/**'
+    - 'packages/**'
+  pull_request:
+    paths:
+    - 'examples/**'
+    - 'packages/**'
+
+env:
+  flutter_channel: 'stable'
+  flutter_version: '1.12.13+hotfix.7'
```
Comment (felangel) [inline .github/workflows/examples.yaml]: can we have a specified `flutter_version` and `flutter_channel` so that we don't need to update it in multiple places (main workflow and examples workflow)?

Comment (tenhobi) [reply .github/workflows/examples.yaml]: Ehh, probably not. I mean we could probably fetch that using bash task inside of each job, but...

File: .github/workflows/examples.yaml (hunk 3 above)
Comment (netantho) [reply .github/workflows/examples.yaml]: Good catch, thanks for the suggestion and snippet @tenhobi!
For `examples.yaml`, I was thinking we may want to run the CI when anything in packages/** gets modified so that we rerun the examples on library changes to make sure the examples are updated, what do you think?

File: packages/angular_bloc/example/example.dart (hunk 4 above)
Comment (netantho) [reply packages/angular_bloc/example/example.dart]: Oh good to know, will revert it and leave a comment then. Thanks Felix!

File: scripts/ci_build_js.sh (hunk 5 above)
Comment (netantho) [reply scripts/ci_build_js.sh]: True, it's not used, thanks Felix!

File: scripts/ci_build_apk.sh (hunk 6 above)
Comment (netantho) [reply scripts/ci_build_apk.sh]: Removed.

File: packages/bloc_test/pubspec.yaml (hunk 7 above)
Comment (netantho) [reply packages/bloc_test/pubspec.yaml]: Oh right, my thinking was to make sure to test cross-packages dependencies breaking changes in pull requests.
Thanks for guiding my through all the rookie mistakes :-)

File: packages/angular_bloc/pubspec.yaml (hunk 8 above)
Comment (netantho) [reply packages/angular_bloc/pubspec.yaml]: Reverted

File: examples/github_search/flutter_github_search/pubspec.yaml (hunk 9 above)
Comment (netantho) [reply examples/github_search/flutter_github_search/pubspec.yaml]: I was hoping it would allow to help not introducing changes in flutter_bloc that would break this example and also reducing the number of places to changes deps when releasing a new version of flutter_bloc. For example, here it was still at 3.0.0 whereas the latest is 3.2.0.
What do you think?

File: .github/workflows/examples.yaml (hunk 10 above)
Comment (netantho) [reply .github/workflows/examples.yaml]: Great suggestion!
`subosito/flutter-action@v1` action uses variables in the github action workflow definition, so we may have to deviate a bit from its interface if adding a bash task.
I don't have strong opinion on this, @felangel do you like adding a bash task option?

File: examples/github_search/flutter_github_search/pubspec.yaml (hunk 9 above)
Comment (netantho) [reply examples/github_search/flutter_github_search/pubspec.yaml]: Actually it's not possible after changing packages to specific versions:

> pub get failed (1; So, because flutter_github_search depends on both flutter_bloc from path and common_github_search from path, version solving failed.)

Putting 3.2.0 specifically.

**Comment by netantho** [comment]
Thanks for the reviews and your patience @tenhobi and @felangel 🙏 👍 💯 
Please have a look at the changes when you have some time :-)

File: .github/workflows/examples.yaml (hunk 3 above)
Comment (tenhobi) [reply .github/workflows/examples.yaml]: As the examples use specific version, nothing should really change. So I would opt out the path.

File: .github/workflows/examples.yaml [hunk 11]
```diff
@@ -0,0 +1,109 @@
+name: examples
+
+on:
+  push:
+    paths:
+    - 'examples/**'
+    - 'packages/**'
```
Comment (tenhobi) [inline .github/workflows/examples.yaml]:
```suggestion
    - 'packages/*/example/**'
    - 'performance_tests/**'
```

File: .github/workflows/examples.yaml [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+  pull_request:
+    paths:
+    - 'examples/**'
+    - 'packages/**'
```
Comment (tenhobi) [inline .github/workflows/examples.yaml]:
```suggestion
    - 'packages/*/example/**'
    - 'performance_tests/**'
```

File: CONTRIBUTING.md [hunk 13]
```diff
@@ -97,6 +97,17 @@ so other contributors can easily see what's already done.
     - [ ] VSCode
 ```
 
+## Adding an example
+
+Examples live in the `examples` folder.
+
+If you're adding an example, make sure to add CI checks for it:
+* For a flutter example: add it to the `flutter_folder` in
```
Comment (tenhobi) [inline CONTRIBUTING.md]:
```suggestion

- For a flutter example, add it to the `flutter_folder` in
```

Nit. We use `-` for lists.

File: CONTRIBUTING.md [hunk 14]
```diff
@@ -97,6 +97,17 @@ so other contributors can easily see what's already done.
...
+[examples.yaml](https://github.com/felangel/bloc/blob/master/.github/workflows/examples.yaml).
+* For a non flutter example (pure dart or angular): add it to the `non_flutter_folder` in
```
Comment (tenhobi) [inline CONTRIBUTING.md]:
```suggestion
- For a non-flutter example (pure dart or angular), add it to the `non_flutter_folder` in
```

Nit. We use `-` for lists.

File: CONTRIBUTING.md [hunk 15]
```diff
@@ -97,6 +97,17 @@ so other contributors can easily see what's already done.
...
+[examples.yaml](https://github.com/felangel/bloc/blob/master/.github/workflows/examples.yaml).
+
+
```
Comment (tenhobi) [inline CONTRIBUTING.md]:
```suggestion
```

File: README.md [hunk 16]
```diff
@@ -1,6 +1,7 @@
 <img src="https://raw.githubusercontent.com/felangel/bloc/master/docs/assets/bloc_logo_full.png" height="60" alt="Bloc" />
 
 [![build](https://github.com/felangel/bloc/workflows/build/badge.svg)](https://github.com/felangel/bloc/actions)
+[![examples](https://github.com/felangel/bloc/workflows/examples/badge.svg)](https://github.com/felangel/bloc/actions)
```
Comment (tenhobi) [inline README.md]: You can also include this badge to each package's README. :wink:

File: .github/workflows/examples.yaml (hunk 3 above)
Comment (netantho) [reply .github/workflows/examples.yaml]: Sounds good, I've whitelisted the package examples from the examples workflow.

File: .github/workflows/examples.yaml (hunk 11 above)
Comment (netantho) [reply .github/workflows/examples.yaml]: The examples in `examples/` use paths for bloc packages dependencies. With the proposed version, we'd check changes in `packages/` with that corpus of examples on each commit and PR. With your suggestion my understanding is that we'd not verify the compatibility of library changes against examples, just modifications to examples themselves.

File: .github/workflows/examples.yaml (hunk 12 above)
Comment (netantho) [reply .github/workflows/examples.yaml]: Same as above

File: README.md (hunk 16 above)
Comment (netantho) [reply README.md]: Great idea! Done

**Comment by netantho** [comment]
@tenhobi @felangel Finished another round of changes based on your feedback. Please have a look you get some time and thanks for the flow of great ideas 💯 !

File: .github/workflows/examples.yaml (hunk 11 above)
Comment (tenhobi) [reply .github/workflows/examples.yaml]: OK, I see.

1) Well, when someone does a change to the package, we should build and test only packages. And not all the examples. (**`packages/**` in main.yaml**)
2) And we should build examples only when there is some change to it. (**`performance_tests/**` and `examples/**` in examples.yaml**)
3) A special case is an example in a package, which is using the bloc etc. reference (`../bloc`). But if we add 'packages/**' to the examples build, we would had to do the testing and build everytime something changes, no matter where. So that would not help, because testing alld `examples/` is time consuming. So I would probably suggest moving `packages/*/example` to main CI? **So that would mean adding example test & build to main.yaml**. What do you think @netantho

File: README.md (hunk 16 above)
Comment (tenhobi) [reply README.md]: @netantho I mean also to packages/{bloc, flutter_bloc, ...}/README.md Next to the main build badge. :-)

image

File: .github/workflows/examples.yaml (hunk 11 above)
Comment (tenhobi) [reply .github/workflows/examples.yaml]: OH, wow. Sorry, this was my bad, sorry for the confusion. I can see now that the examples also use relative path.

For example
```
flutter_bloc:
  path: ../../packages/flutter_bloc
``` 

I thought they are fixed to a specific version. OK, in that case, I guess we have to build everything on any change to the repository?

Oh, I can see that those builds take surprisingly not so much time, so that would be OK I guess? Cc @felangel 

For example:
image

Comment (tenhobi) [reply .github/workflows/examples.yaml]: In that case, we actually could merge those CI files? And create jobs like `packages` (current main.yaml) and `examples`. Simple names, so we don't have to have something crazy like `analyze-and-test-and-build-apk`. 🤣

Comment (tenhobi) [reply .github/workflows/examples.yaml]: And please, just a nit: could you reorder matrix so folders are first and OS second? That would help to create better previews, now it shows OS first and that not useful. 😄 💙

Comment (netantho) [reply .github/workflows/examples.yaml]: Thanks @tenhobi! Implemented 👍

File: .github/workflows/main.yaml [hunk 17]
```diff
@@ -1,20 +1,30 @@
 name: build
 
-on: [push, pull_request]
+on:
+  push:
+    paths:
+    - 'packages/**'
+    - '!packages/*/example/**'
```
Comment (tenhobi) [inline .github/workflows/main.yaml]:
```suggestion
    - 'performance_tests/**'
```

We don't need this anymore. Plus we forgot including this folder. :)

Protip: you can just click in the PR "accept suggestion" to speed up the process. ;)

File: .github/workflows/main.yaml [hunk 18]
```diff
@@ -1,20 +1,30 @@
...
+    - 'examples/**'
+  pull_request:
+    paths:
+    - 'packages/**'
+    - '!packages/*/example/**'
```
Comment (tenhobi) [inline .github/workflows/main.yaml]:
```suggestion
    - 'performance_tests/**'
```

We don't need this anymore. Plus we forgot including this folder. :)

Protip: you can just click in the PR "accept suggestion" to speed up the process. ;)

File: CONTRIBUTING.md [hunk 19]
```diff
@@ -97,6 +97,16 @@ so other contributors can easily see what's already done.
     - [ ] VSCode
 ```
 
+## Adding an example
+
+Examples live in the `examples` folder.
+
+If you're adding an example, make sure to add CI checks for it:
+- For a flutter example: add it to the `flutter_folder` in
+[examples.yaml](https://github.com/felangel/bloc/blob/master/.github/workflows/examples.yaml).
```
Comment (tenhobi) [inline CONTRIBUTING.md]:
```suggestion
[main.yaml](https://github.com/felangel/bloc/blob/master/.github/workflows/main.yaml).
```

File: CONTRIBUTING.md [hunk 20]
```diff
@@ -97,6 +97,16 @@ so other contributors can easily see what's already done.
...
+- For a non flutter example (pure dart or angular): add it to the `non_flutter_folder` in
+[examples.yaml](https://github.com/felangel/bloc/blob/master/.github/workflows/examples.yaml).
```
Comment (tenhobi) [inline CONTRIBUTING.md]:
```suggestion
[main.yaml](https://github.com/felangel/bloc/blob/master/.github/workflows/main.yaml).
```

File: README.md (hunk 16 above)
Comment (tenhobi) [inline README.md]:
```suggestion
```

We merged the builds, so we have only build badge.

File: examples/angular_counter/README.md [hunk 21]
```diff
@@ -1,3 +1,5 @@
+[![build](https://github.com/felangel/bloc/workflows/build/badge.svg)](https://github.com/felangel/bloc/actions)
```
Comment (tenhobi) [inline examples/angular_counter/README.md]:
```suggestion
```

We merged the builds, so we have only build badge.

File: examples/flutter_form_validation/README.md [hunk 22]
```diff
@@ -1,3 +1,5 @@
+[![examples](https://github.com/felangel/bloc/workflows/examples/badge.svg)](https://github.com/felangel/bloc/actions)
```
Comment (tenhobi) [inline examples/flutter_form_validation/README.md]:
```suggestion
[![build](https://github.com/felangel/bloc/workflows/build/badge.svg)](https://github.com/felangel/bloc/actions)
```

Comment (netantho) [reply examples/flutter_form_validation/README.md]: Oops, good catch!

**Comment by netantho** [comment]
Many thanks for the review 👍 💯 ! Hopefully I didn't miss anything this time 🙏

**Comment by felangel** [comment]
Y'all are awesome! 💯  I'll review this PR first thing tomorrow 😴

File: .github/workflows/main.yaml (hunk 18 above)
Comment (tenhobi) [reply .github/workflows/main.yaml]: @netantho this is still wrong. Btw. if you would have use the suggestions and accepted them, it would work out. :P

Comment (netantho) [reply .github/workflows/main.yaml]: @tenhobi I'm very sorry, I read too fast, got confused with another comment. 🤦‍♂  Thank you very much for checking!

**Comment by netantho** [comment]
> Still missing the fix in the path for pull request, see the comment.

I'm so sorry, read too fast too early in the morning and got confused with another comment. I hope that https://github.com/felangel/bloc/pull/822/commits/e5e83652c11bc5a3cdcc1cd56cb16ec6ac9b73e5 solves it.

File: .github/workflows/main.yaml [hunk 23]
```diff
@@ -41,10 +51,102 @@ jobs:
...
+          export PATH=$PATH:$FLUTTER_DART_HOME
+          ./scripts/ci_build_apk.sh ${{ matrix.folder }}
+
+  # Analyze, test and build for each web example.
+  examples-web:
+    runs-on: ${{ matrix.os }}
+
+    strategy:
+      matrix:
+        folder: ['packages/bloc/example', 'packages/bloc_test/example', 'examples/angular_counter', 'examples/github_search/common_github_search', 'examples/github_search/angular_github_search']
```
Comment (felangel) [inline .github/workflows/main.yaml]: why is the `bloc/example` and `bloc_test/example` included in `examples-web`? They are pure dart examples.

**Comment by felangel** [comment]
> Hi, very nice! I was thinking about the same but didn't have time to do it. :-)
> 
> We should consider creating its own action (`examples.yaml`) so it's not run every time when nothing changed there. That means copying exactly the `main.yaml`, but also adding path settings. Otherwise, any change would also check those examples, which would take no time + have no impact with not changed settings.
> 
> Take a look at https://help.github.com/en/actions/automating-your-workflow-with-github-actions/workflow-syntax-for-github-actions#onpushpull_requestpaths
> 
> Edit: And we could then make separate badge `examples: passing`. 😄

I agree with this. @netantho thoughts on separating this into a separate workflow?

File: examples/flutter_bloc_with_stream/lib/bloc/ticker_bloc.dart [hunk 24]
```diff
@@ -25,7 +25,7 @@ class TickerBloc extends Bloc<TickerEvent, TickerState> {
 
   @override
   Future<void> close() {
-    subscription.cancel();
+    subscription?.cancel();
```
Comment (felangel) [inline examples/flutter_bloc_with_stream/lib/bloc/ticker_bloc.dart]: 💯

File: packages/angular_bloc/example/example.dart [hunk 25]
```diff
@@ -1,69 +1,71 @@
+// This example is one-pager for the [angular_bloc](https://github.com/felangel/bloc/tree/master/examples/angular_counter) example.
+
 import 'dart:async';
 
-import 'package:angular/angular.dart';
+import 'package:angular/angular.dart';
```
Comment (felangel) [inline packages/angular_bloc/example/example.dart]: can we revert all this additional whitespace that was added?

File: .github/workflows/main.yaml (hunk 23 above)
Comment (netantho) [reply .github/workflows/main.yaml]: That's true that we don't do anything more than on `packages/*` here, we just do a pub get and then analyze.
Removing these folders.

**Comment by netantho** [comment]
> > Hi, very nice! I was thinking about the same but didn't have time to do it. :-)
> > We should consider creating its own action (`examples.yaml`) so it's not run every time when nothing changed there. That means copying exactly the `main.yaml`, but also adding path settings. Otherwise, any change would also check those examples, which would take no time + have no impact with not changed settings.
> > Take a look at https://help.github.com/en/actions/automating-your-workflow-with-github-actions/workflow-syntax-for-github-actions#onpushpull_requestpaths
> > Edit: And we could then make separate badge `examples: passing`. 😄
> 
> I agree with this. @netantho thoughts on separating this into a separate workflow?

We had an iteration with separate workflows and then rolled back.
The main reason for a single workflow is that examples have relative path deps to packages. So when a package gets modified we want to run the examples checks. If we separate into two workflows, both will be run when there's a change in `packages/**` which could be confusing.
Also with a single workflow we can easily share `flutter_channel` and `flutter_version` env variables.

@tenhobi if you want add some pros/cons :-)

File: examples/flutter_bloc_with_stream/lib/bloc/ticker_bloc.dart (hunk 24 above)
Comment (netantho) [reply examples/flutter_bloc_with_stream/lib/bloc/ticker_bloc.dart]: A bug found with the CI ;-)

File: packages/angular_bloc/example/example.dart (hunk 25 above)
Comment (netantho) [reply packages/angular_bloc/example/example.dart]: Not sure what happened in there 😲 Whitespace party? 🎉 
Thanks!

**Comment by tenhobi** [comment]
As @netantho said. At first I have thought that each package has its version of packages locked by a version. But I have later found that it is not that case as it used relative paths, which is ok. So any change to packages might break examples and therefore we have to check them too.

File: .github/workflows/main.yaml (hunk 23 above)
Comment (tenhobi) [reply .github/workflows/main.yaml]: I am not sure removing these folders is a good way. More like creating examples-pure that contains only analysis? It seems to me like you have deleted it.

**Comment by felangel** [comment]
Thanks so much for your hard work! This is really amazing and I really appreciate it 💙
